src/fs/read_dir.rs

```diff
@@ -57,25 +64,57 @@ pub async fn read_dir<'js>(
     path: String,
     options: Opt<Object<'js>>,
 ) -> Result<ReadDir> {
+    let mut path = path;
     let with_file_types = options
         .0
+        .as_ref()
         .and_then(|opts| opts.get("withFileTypes").ok())
         .and_then(|file_types: Value| file_types.as_bool())
         .unwrap_or_default();
 
-    let mut dir = fs::read_dir(path).await.or_throw(&ctx)?;
+    let with_recursive = options
+        .0
+        .as_ref()
+        .and_then(|opts| opts.get("recursive").ok())
+        .and_then(|recursive: Value| recursive.as_bool())
+        .unwrap_or_default();
 
+    let skip_root_pos = {
+        match path.as_str() {
+            // . | ./
+            "." | CURRENT_DIR_STR => CURRENT_DIR_STR.len(),
+            // ./path
+            _ if path.starts_with(CURRENT_DIR_STR) => path.len() + 1,
+            // path
+            _ => {
+                if !is_absolute(path.clone()) {
+                    path = [CURRENT_DIR_STR.to_string(), path].concat();
+                }
+                path.len() + 1
+            }
+        }
+    };
     let mut items = Vec::with_capacity(64);
+    let mut dirs = VecDeque::from_iter([PathBuf::from(path)]);
 
-    while let Some(child) = dir.next_entry().await? {
-        if let Some(name) = child.path().file_name() {
-            let name = name.to_string_lossy().to_string();
+    while let Some(mut dir_buf) = dirs.pop_back() {
```

Comment:
Nice we have a non-recursive traversal 👍 
I have similar functionality in here that could be changed to return an iterator instead and reused here.
https://github.com/awslabs/llrt/blob/37eb92b062f858a42567c1457ee199d977c51153/src/utils/io.rs#L45
Also, we might speed up things if we parallelize the walk of subdirectories by spawning tasks that joining the awaits that calls next entries.

Reply:
Thanks for your suggestion, I added `DirectoryWalker` to replace them, you can see if they meet your expectations